### PR TITLE
Added FarmlandTrampleEvent

### PR DESCRIPTION
**Pull request by minzmann**
An event that can be canceled to prevent farmland from being trampled.

This makes it possible to: making certain plants resist trampling, preventing players from trampling under certain circumstances, disable drops if plants are trampled (by setting the block/s above to air), causing certain effects,...

Changes to base classes is kept minimal (one line)

### CODE REVIEW

**Comment by cabralRodrigo** [comment]
+1 :+1:

**Comment by Spartan322** [comment]
Not sure for this type of patch, but I think you're suppose to use ForgeHooks still.

**Comment by Lunatrius** [comment]
It works without `ForgeHooks`, but should be moved for the sake of readability. That aside, you can make the patch smaller by adding the if to the previous line.

```patch
-                worldIn.setBlockState(pos, Blocks.dirt.getDefaultState());
+                if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.FarmlandTrampleEvent(worldIn, pos, worldIn.getBlockState(pos), (EntityLivingBase) entityIn))) worldIn.setBlockState(pos, Blocks.dirt.getDefaultState());
```
becomes
```patch
+                if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.FarmlandTrampleEvent(worldIn, pos, worldIn.getBlockState(pos), (EntityLivingBase) entityIn))) // Don't indent to lower patch size
                 worldIn.setBlockState(pos, Blocks.dirt.getDefaultState());
```

**Comment by minzmann** [comment]
Updated to include suggested changes

File: patches/minecraft/net/minecraft/block/BlockFarmland.java.patch [hunk 1]
```diff
@@ -1,6 +1,14 @@
 --- ../src-base/minecraft/net/minecraft/block/BlockFarmland.java
 +++ ../src-work/minecraft/net/minecraft/block/BlockFarmland.java
-@@ -89,7 +89,7 @@
+@@ -79,6 +79,7 @@
+                     return;
+                 }
+ 
++                if (!net.minecraftforge.common.ForgeHooks.onFarmlandTrample(p_180658_1_, p_180658_2_, (EntityLivingBase) p_180658_3_))
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/block/BlockFarmland.java.patch]: no space after cast

File: src/main/java/net/minecraftforge/event/world/FarmlandTrampleEvent.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Called just before farmland would change to dirt.
+ * If the event is cancelled, the farmland (and plant) will remain intact.
+ * <br>
+ * This event is {@link Cancelable}.<br>
+ * This event does not use {@link HasResult}.<br>
+ * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+ * 
+ * @param pos The position of the farmland block
+ * @param entity The entity that trampled the farmland
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/event/world/FarmlandTrampleEvent.java]: Params go on constructors not classes....

**Comment by LexManos** [comment]
Don't we already have a block change events thats captured to satisfy this?
Also events should use getters/setters for values now, it's a rule for new events. {We'll eventually convert the old events}
Also this needs more context, modders should be able to expand this to whatever blocks they have. Custom farmlands ftw.
Also it may be worth it to converting it into a return of IBlockState so the event can change the result.
For example:
Sheep trampling dry farmland = dirt
Humans trampling wet farmland = clay.

**Comment by minzmann** [comment]
> Don't we already have a block change events thats captured to satisfy this?

Not really, the closest thing to a block change event ist NeighborNotifyEvent and it would seem kind of weird adding this functionality to that event

> Also this needs more context, modders should be able to expand this to whatever blocks they have. 

Whats stopping them (except that it hasn't been merged yet)? It takes a generic position and state, nothing that limits it to vanilla farmland

> Also it may be worth it to converting it into a return of IBlockState so the event can change the result.

I'm not seeing the huge benefit of returning the state compared to just calling world.setBlockState inside the event handler and cancelling the event

**Comment by LexManos** [comment]
Changing the return allows for multiple listeners to interact with it and know what the state is going to be.
Canceling it and setting the block itself bypasses the handlers that just want to cancel it entirely.

**Comment by minzmann** [comment]
The event now has an additional variable storing the target state (and a more generic name). After the event has been processed the block's state is updated to the target one (if the event hasn't been canceled).

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Comment by bs2609** [comment]
@minzmann If you're still interested, feel like retargeting this to 1.10.x branch?

**Pull request closed without merging**